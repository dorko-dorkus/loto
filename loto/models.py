"""Core data models for lock-out/tag-out planning utilities.

This module defines strict Pydantic models used throughout the project.  All
models forbid extra fields to help surface typing mistakes early and include
basic metadata/units for numeric values.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class DomainRule(BaseModel):
    """Represents a rule describing domain-specific constraints."""

    name: str = Field(..., description="Rule name")
    expression: str = Field(..., description="Expression used to evaluate the rule")

    class Config:
        extra = "forbid"


class VerificationRule(BaseModel):
    """Represents a rule verifying an expected condition."""

    name: str = Field(..., description="Verification rule name")
    check: str = Field(..., description="Expression or callable for verification")

    class Config:
        extra = "forbid"


class RiskPolicies(BaseModel):
    """Policy thresholds for different risk levels.

    The ``levels`` mapping associates a textual risk level with a probability
    threshold in the range ``[0, 1]`` (unit: probability).
    """

    levels: Dict[str, float] = Field(
        default_factory=dict,
        description="Mapping of risk level to probability threshold [0-1]",
    )

    class Config:
        extra = "forbid"


class Node(BaseModel):
    """A node within a graph structure."""

    id: str = Field(..., description="Unique node identifier")
    label: Optional[str] = Field(None, description="Human readable label")

    class Config:
        extra = "forbid"


class Edge(BaseModel):
    """An edge connecting two nodes within a graph."""

    source: str = Field(..., description="Identifier of the source node")
    target: str = Field(..., description="Identifier of the target node")
    weight: Optional[float] = Field(
        None, description="Edge weight (unitless)")

    class Config:
        extra = "forbid"


class GraphBundle(BaseModel):
    """Group of graph components accompanied by metadata."""

    nodes: List[Node] = Field(
        default_factory=list, description="Nodes comprising the graph"
    )
    edges: List[Edge] = Field(
        default_factory=list, description="Edges comprising the graph"
    )
    metadata: Dict[str, Any] = Field(
        default_factory=dict, description="Arbitrary graph metadata"
    )

    class Config:
        extra = "forbid"


class IsolationAction(BaseModel):
    """Action executed to isolate a component."""

    component_id: str = Field(..., description="Identifier of the component")
    method: str = Field(
        ..., description="Isolation method such as 'lock' or 'tag'"
    )
    duration_s: Optional[float] = Field(
        None, description="Expected duration in seconds"
    )

    class Config:
        extra = "forbid"


class IsolationPlan(BaseModel):
    """Plan made up of a sequence of isolation actions."""

    plan_id: str = Field(..., description="Unique identifier for the plan")
    actions: List[IsolationAction] = Field(
        default_factory=list, description="Ordered isolation actions"
    )

    class Config:
        extra = "forbid"


class Stimulus(BaseModel):
    """Input stimulus used for simulation."""

    name: str = Field(..., description="Stimulus name")
    magnitude: float = Field(..., description="Stimulus magnitude (unitless)")
    duration_s: float = Field(..., description="Stimulus duration in seconds")

    class Config:
        extra = "forbid"


class SimResultItem(BaseModel):
    """Result produced from simulating a single stimulus."""

    stimulus: Stimulus = Field(..., description="Stimulus that was simulated")
    success: bool = Field(..., description="Whether the simulation succeeded")
    impact: float = Field(..., description="Impact score (unitless)")

    class Config:
        extra = "forbid"


class SimReport(BaseModel):
    """Collection of simulation results."""

    results: List[SimResultItem] = Field(
        default_factory=list, description="Individual simulation results"
    )
    total_time_s: float = Field(
        ..., description="Total simulation time in seconds"
    )

    class Config:
        extra = "forbid"


class ImpactReport(BaseModel):
    """Quantified impact report for a component or system."""

    component_id: str = Field(..., description="Component identifier")
    severity: float = Field(..., description="Severity score (unitless)")
    description: Optional[str] = Field(
        None, description="Human readable description of the impact"
    )

    class Config:
        extra = "forbid"


class ArtifactBundle(BaseModel):
    """Bundle of generated artifacts."""

    artifacts: Dict[str, str] = Field(
        default_factory=dict,
        description="Mapping of artifact name to path or URI",
    )
    metadata: Dict[str, Any] = Field(
        default_factory=dict, description="Additional artifact metadata"
    )

    class Config:
        extra = "forbid"


class RulePack(BaseModel):
    """Collection of domain and verification rules with optional risk policies."""

    domain_rules: List[DomainRule] = Field(
        default_factory=list, description="Rules describing domain constraints"
    )
    verification_rules: List[VerificationRule] = Field(
        default_factory=list, description="Rules verifying expected conditions"
    )
    risk_policies: Optional[RiskPolicies] = Field(
        None, description="Associated risk policies"
    )

    class Config:
        extra = "forbid"
