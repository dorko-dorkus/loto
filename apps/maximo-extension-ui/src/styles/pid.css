.pid-container {
  position: relative;
  width: 100%;
  height: 100%;
}

.pid-container:focus-visible {
  outline: 2px solid var(--mxc-topbar-bg);
  outline-offset: 2px;
}

.hl-primary {
  stroke: var(--mxc-topbar-bg);
  stroke-width: 2;
}

.hl-warning {
  stroke: #fbbf24;
  stroke-width: 2;
}

.badge {
  background: var(--mxc-topbar-bg);
  color: var(--mxc-topbar-fg);
  padding: 2px 6px;
  border-radius: var(--mxc-radius-sm);
}

.pid-badge {
  display: inline-block;
  padding: 2px 4px;
  border-radius: var(--mxc-radius-sm);
  font-size: 0.7rem;
  color: #fff;
}

.pid-badge-asset {
  background: var(--mxc-topbar-bg);
}

.pid-badge-source {
  background: #3b82f6;
}

.pid-badge-missing-tag {
  background: #f87171;
}

.pid-badge-unknown-selector {
  background: #f97316;
}

.pid-badge-unmapped {
  background: #9ca3af;
}

.pid-warning-drawer {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  max-width: 12rem;
  background: white;
  padding: 0.5rem;
  border-radius: var(--mxc-radius-sm);
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
}

.pid-warning-chip {
  display: inline-block;
  background: #fbbf24;
  color: #000;
  padding: 2px 4px;
  border-radius: var(--mxc-radius-sm);
  margin: 0 4px 4px 0;
  font-size: 0.75rem;
  border: none;
  cursor: pointer;
}
