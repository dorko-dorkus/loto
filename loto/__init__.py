"""Top-level package for the LOTO planner.

This package contains the core components needed to compute lockout–tagout
plans, run simulation tests, and integrate with external systems like
Maximo, WAPR and Coupa. Only class and method stubs are defined here.

Each module exposes a primary class responsible for a distinct concern:

* :mod:`loto.rule_engine` – Loading and validating rule packs.
* :mod:`loto.graph_builder` – Building connectivity graphs from raw data.
* :mod:`loto.isolation_planner` – Computing isolation plans using cut-set algorithms.
* :mod:`loto.sim_engine` – Applying isolation plans and running stimulus tests.
* :mod:`loto.integrations` – Adapters for external systems (CMMS, permit, procurement).
* :mod:`loto.renderer` – Generating human‐readable plan documents.
* :mod:`loto.cli` – Command line entry points (not implemented yet).
"""

__all__ = [
    "RuleEngine",
    "GraphBuilder",
    "IsolationPlanner",
    "SimEngine",
    "IntegrationAdapter",
    "Renderer",
]

# Import statements are deferred until runtime to avoid circular imports
