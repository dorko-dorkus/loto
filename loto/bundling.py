"""Utilities for selecting candidate work bundles.

This module implements a simple 0/1 knapsack style selector used to pick
maintenance candidates that maximise saved future derate subject to two
resource constraints: readiness and SIMOPs (simultaneous operations).

The function :func:`select_candidates` returns both the chosen subset and
human readable reasons for why each candidate was or was not selected.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class Candidate:
    """A potential maintenance task to be considered for bundling."""

    name: str
    saved_future_derate: float
    readiness_cost: int
    simops_cost: int
    ready: bool = True


ReasonMap = dict[str, str]


def _knapsack(
    candidates: Sequence[Candidate], max_readiness: int, max_simops: int
) -> set[int]:
    """Return indices of the optimal candidate subset.

    A dynamic programming approach is used to solve a two dimensional
    knapsack problem where readiness and SIMOPs represent the capacity
    constraints.
    """

    # dp[r][s] -> (value, set_of_indices)
    dp: list[list[tuple[float, set[int]]]] = [
        [(0.0, set()) for _ in range(max_simops + 1)] for _ in range(max_readiness + 1)
    ]

    for idx, c in enumerate(candidates):
        for r in range(max_readiness, c.readiness_cost - 1, -1):
            for s in range(max_simops, c.simops_cost - 1, -1):
                prev_val, prev_set = dp[r - c.readiness_cost][s - c.simops_cost]
                new_val = prev_val + c.saved_future_derate
                if new_val > dp[r][s][0]:
                    dp[r][s] = (new_val, prev_set | {idx})

    return dp[max_readiness][max_simops][1]


def select_candidates(
    candidates: Iterable[Candidate],
    max_readiness: int,
    max_simops: int,
) -> tuple[list[Candidate], ReasonMap]:
    """Select candidates that maximise saved future derate.

    Parameters
    ----------
    candidates:
        Iterable of :class:`Candidate` objects to consider.
    max_readiness:
        Maximum total readiness cost available.
    max_simops:
        Maximum total SIMOPs cost allowed.

    Returns
    -------
    tuple[list[Candidate], dict[str, str]]
        The chosen subset of candidates and a mapping from candidate name
        to a human readable reason for inclusion or exclusion.
    """

    candidates = list(candidates)
    reasons: ReasonMap = {}

    # Filter out candidates that are not ready
    ready_candidates: list[Candidate] = []
    for c in candidates:
        if not c.ready:
            reasons[c.name] = "not ready"
        else:
            ready_candidates.append(c)

    chosen_indices = _knapsack(ready_candidates, max_readiness, max_simops)
    selected: list[Candidate] = []
    for idx, c in enumerate(ready_candidates):
        if idx in chosen_indices:
            selected.append(c)
            reasons[c.name] = f"selected ({c.saved_future_derate} MW saved)"

    # Any remaining ready candidate was excluded due to constraints
    for c in ready_candidates:
        if c.name not in reasons:
            if c.readiness_cost > max_readiness or c.simops_cost > max_simops:
                reasons[c.name] = "exceeds individual constraints"
            else:
                reasons[c.name] = "excluded to respect readiness/SIMOPs constraints"

    return selected, reasons
