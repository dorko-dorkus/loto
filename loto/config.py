from __future__ import annotations

from dataclasses import dataclass, field
import os
from pathlib import Path
from typing import Dict, Optional

from dotenv import load_dotenv


class ConfigError(Exception):
    """Exception raised when configuration is invalid.

    Attributes
    ----------
    code: str
        Machine readable error code.
    hint: str
        Human readable hint describing how to fix the issue.
    """

    def __init__(self, *, code: str, hint: str) -> None:
        self.code = code
        self.hint = hint
        super().__init__(hint)


@dataclass
class MaximoConfig:
    mode: str
    base_url: Optional[str] = None
    apikey: Optional[str] = None
    os: Dict[str, str] = field(default_factory=dict)


@dataclass
class IntegrationConfig:
    mode: str
    base_url: Optional[str] = None
    apikey: Optional[str] = None


@dataclass
class AppConfig:
    app_env: str
    data_in: str
    data_out: str
    rulepack_file: str
    maximo: MaximoConfig
    wapr: IntegrationConfig
    coupa: IntegrationConfig


CONFIG_ERROR_CODE = "CONFIG/ENV"


def _required(keys: list[str]) -> None:
    missing = [k for k in keys if not os.getenv(k)]
    if missing:
        raise ConfigError(code=CONFIG_ERROR_CODE, hint=f"Missing environment variables: {', '.join(missing)}")


def load_config() -> AppConfig:
    """Load configuration from ``.env.<APP_ENV>`` and environment variables."""

    app_env = os.getenv("APP_ENV", "demo").lower()
    if app_env not in {"demo", "live"}:
        raise ConfigError(code=CONFIG_ERROR_CODE, hint="APP_ENV must be 'demo' or 'live'")

    # load .env file for the selected environment if present
    load_dotenv(Path(f".env.{app_env}"), override=False)

    # basic application settings
    _required(["DATA_IN", "DATA_OUT", "RULEPACK_FILE"])
    data_in = os.getenv("DATA_IN")
    data_out = os.getenv("DATA_OUT")
    rulepack_file = os.getenv("RULEPACK_FILE")

    # maximo settings
    maximo_mode = os.getenv("MAXIMO_MODE", "MOCK").upper()
    maximo_base_url = os.getenv("MAXIMO_BASE_URL")
    maximo_apikey = os.getenv("MAXIMO_APIKEY")
    maximo_os = {
        k.removeprefix("MAXIMO_OS_"): v
        for k, v in os.environ.items()
        if k.startswith("MAXIMO_OS_")
    }
    if maximo_mode == "LIVE":
        missing = []
        if not maximo_base_url:
            missing.append("MAXIMO_BASE_URL")
        if not maximo_apikey:
            missing.append("MAXIMO_APIKEY")
        if not maximo_os:
            missing.append("MAXIMO_OS_*")
        if missing:
            raise ConfigError(
                code=CONFIG_ERROR_CODE,
                hint="Missing Maximo configuration: " + ", ".join(missing),
            )
    maximo = MaximoConfig(
        mode=maximo_mode,
        base_url=maximo_base_url,
        apikey=maximo_apikey,
        os=maximo_os,
    )

    # wapr settings
    wapr_mode = os.getenv("WAPR_MODE", "MOCK").upper()
    wapr_base_url = os.getenv("WAPR_BASE_URL")
    wapr_apikey = os.getenv("WAPR_APIKEY")
    if wapr_mode == "LIVE":
        missing = []
        if not wapr_base_url:
            missing.append("WAPR_BASE_URL")
        if not wapr_apikey:
            missing.append("WAPR_APIKEY")
        if missing:
            raise ConfigError(
                code=CONFIG_ERROR_CODE,
                hint="Missing WAPR configuration: " + ", ".join(missing),
            )
    wapr = IntegrationConfig(mode=wapr_mode, base_url=wapr_base_url, apikey=wapr_apikey)

    # coupa settings
    coupa_mode = os.getenv("COUPA_MODE", "MOCK").upper()
    coupa_base_url = os.getenv("COUPA_BASE_URL")
    coupa_apikey = os.getenv("COUPA_APIKEY")
    if coupa_mode == "LIVE":
        missing = []
        if not coupa_base_url:
            missing.append("COUPA_BASE_URL")
        if not coupa_apikey:
            missing.append("COUPA_APIKEY")
        if missing:
            raise ConfigError(
                code=CONFIG_ERROR_CODE,
                hint="Missing Coupa configuration: " + ", ".join(missing),
            )
    coupa = IntegrationConfig(mode=coupa_mode, base_url=coupa_base_url, apikey=coupa_apikey)

    return AppConfig(
        app_env=app_env,
        data_in=data_in,  # type: ignore[arg-type]
        data_out=data_out,  # type: ignore[arg-type]
        rulepack_file=rulepack_file,  # type: ignore[arg-type]
        maximo=maximo,
        wapr=wapr,
        coupa=coupa,
    )
