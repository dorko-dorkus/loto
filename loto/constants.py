"""Project-wide constants."""

DOC_CATEGORY = "Permit/LOTO"
DOC_CATEGORY_DIR = DOC_CATEGORY.replace("/", "_")

# Checklist item verifying the permit has been closed and uploaded.
CHECKLIST_HAND_BACK = "Permit Closed & Hand-back uploaded"
