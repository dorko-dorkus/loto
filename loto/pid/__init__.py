"""PID overlay utilities."""

from .overlay import build_overlay

__all__ = ["build_overlay"]
