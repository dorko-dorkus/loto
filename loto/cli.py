"""Command line entry points for the LOTO planner.

This module defines a basic command line interface for running the LOTO
planner. It parses arguments, orchestrates the loading of rules,
building graphs, computing isolation plans, running simulations, and
producing outputs. Only a skeleton structure is provided here.
"""

from __future__ import annotations

import argparse
import json
import shutil
from pathlib import Path
from typing import Optional
from uuid import uuid4

import networkx as nx
import typer
import yaml
from tqdm import tqdm

from .constants import DOC_CATEGORY_DIR
from .graph_builder import GraphBuilder
from .isolation_planner import IsolationPlanner
from .models import SimReport, Stimulus
from .renderer import Renderer
from .rule_engine import RuleEngine
from .sim_engine import SimEngine

cli = typer.Typer()


@cli.callback()
def main_callback() -> None:
    """LOTO command line interface."""
    return None


def parse_args(args: Optional[list[str]] = None) -> argparse.Namespace:
    """Parse command line arguments.

    Parameters
    ----------
    args: Optional[list[str]]
        Arguments to parse, defaults to None which uses sys.argv.

    Returns
    -------
    argparse.Namespace
        The parsed arguments.
    """
    parser = argparse.ArgumentParser(description="LOTO planner CLI")
    parser.add_argument("--asset", help="Asset tag to isolate")
    parser.add_argument("--rules", help="Path to rule pack file (YAML/JSON)")
    parser.add_argument("--line-list", help="Path to line list CSV")
    parser.add_argument("--valves", help="Path to valves CSV")
    parser.add_argument("--drains", help="Path to drains/vents CSV")
    parser.add_argument("--sources", help="Path to energy sources CSV", default=None)
    parser.add_argument(
        "--air-map", help="Path to instrument air map CSV", default=None
    )
    parser.add_argument("--hazards", help="Path to hazards file", default=None)
    parser.add_argument("--controls", help="Path to controls file", default=None)
    parser.add_argument("--output", help="Directory for outputs", default="./out")
    parser.add_argument("--no-sim", action="store_true", help="Skip simulation testing")
    parser.add_argument("--demo", action="store_true", help="Run in demo mode")

    parsed = parser.parse_args(args)

    if not parsed.demo:
        required = ["asset", "rules", "line_list", "valves", "drains"]
        missing = [opt for opt in required if getattr(parsed, opt) is None]
        if missing:
            parser.error(
                "Missing required arguments: "
                + ", ".join("--" + m.replace("_", "-") for m in missing)
            )

    return parsed


def main(argv: Optional[list[str]] = None) -> None:
    """Entry point for the CLI.

    This function orchestrates the high-level flow: load rules, build
    graphs, compute the isolation plan, optionally run simulations,
    and render outputs. Method bodies are left as placeholders.
    """
    args = parse_args(argv)

    if args.demo:
        demo_dir = Path(__file__).resolve().parent.parent / "demo"
        args.asset = args.asset or "A"
        args.rules = args.rules or str(demo_dir / "rules.yaml")
        args.line_list = args.line_list or str(demo_dir / "line_list.csv")
        args.valves = args.valves or str(demo_dir / "valves.csv")
        args.drains = args.drains or str(demo_dir / "drains.csv")
        args.sources = args.sources or str(demo_dir / "sources.csv")
        args.hazards = args.hazards or str(demo_dir / "hazards.yaml")
        args.controls = args.controls or str(demo_dir / "controls.yaml")

    # Instantiate engine components
    rule_engine = RuleEngine()
    graph_builder = GraphBuilder()
    planner = IsolationPlanner()
    sim = SimEngine()
    renderer = Renderer()

    # Load rule pack, falling back to an empty pack if necessary
    try:
        rule_pack = rule_engine.load(args.rules)
    except Exception:  # pragma: no cover - graceful fallback
        from .models import RulePack

        # Explicitly set default fields to satisfy type checkers
        rule_pack = RulePack(domain_rules=[], verification_rules=[], risk_policies=None)

    # Build graphs from CSVs.  The builder is currently unimplemented, so fall
    # back to a minimal graph that contains a single source feeding the asset.
    try:
        graphs = graph_builder.from_csvs(
            args.line_list,
            args.valves,
            args.drains,
            args.sources,
            args.air_map,
        )
        if args.demo:
            for g in graphs.values():
                for u, v, data in g.edges(data=True):
                    if g.nodes[u].get("is_isolation_point") or g.nodes[v].get(
                        "is_isolation_point"
                    ):
                        data["is_isolation_point"] = True
    except Exception:  # pragma: no cover - use dummy graphs when builder missing
        g = nx.MultiDiGraph()
        g.add_node("source", is_source=True)
        g.add_node(args.asset, tag=args.asset)
        g.add_edge("source", args.asset, is_isolation_point=True)
        graphs = {"default": g}

    # Compute the isolation plan
    plan = planner.compute(graphs, args.asset, rule_pack)

    # Load hazards and controls if provided
    if args.hazards:
        try:
            with Path(args.hazards).open("r", encoding="utf-8") as f:
                data = yaml.safe_load(f) or []
                if isinstance(data, list):
                    plan.hazards.extend(str(item) for item in data)
        except Exception:  # pragma: no cover - ignore loading failures
            pass
    if args.controls:
        try:
            with Path(args.controls).open("r", encoding="utf-8") as f:
                data = yaml.safe_load(f) or []
                if isinstance(data, list):
                    plan.controls.extend(str(item) for item in data)
        except Exception:  # pragma: no cover - ignore loading failures
            pass

    # Run simulation if not skipped.  If the simulation engine is not
    # implemented, keep an empty report.
    sim_report: SimReport = SimReport(results=[], total_time_s=0.0, seed=None)
    if not args.no_sim:
        try:
            applied_graphs = sim.apply(plan, graphs)
            default_stimuli = [
                Stimulus(name="REMOTE_OPEN", magnitude=1.0, duration_s=1.0),
                Stimulus(name="LOCAL_OPEN", magnitude=1.0, duration_s=1.0),
            ]
            sim_report = sim.run_stimuli(applied_graphs, default_stimuli, rule_pack)
        except Exception:  # pragma: no cover - ignore simulation failures
            pass

    # Render outputs
    json_output = renderer.to_json(plan, sim_report)
    json_output["category"] = "Permit/LOTO"
    rule_hash = rule_engine.hash(rule_pack)
    pdf_bytes = renderer.pdf(plan, sim_report, rule_hash, seed=None, timezone="UTC")

    out_dir = Path(args.output)
    out_dir.mkdir(parents=True, exist_ok=True)
    json_path = out_dir / f"LOTO_{args.asset}.json"
    with json_path.open("w", encoding="utf-8") as f:
        json.dump(json_output, f, indent=2)

    pdf_path = out_dir / f"LOTO_{args.asset}.pdf"
    with pdf_path.open("wb") as f:
        f.write(pdf_bytes)

    return None


@cli.command()
def demo(out: Path = Path("./out"), open_pdf: bool = False) -> None:
    """Generate demo PDF and JSON outputs."""
    try:
        with tqdm(total=1, desc="Generating demo", unit="step") as progress:
            main(["--demo", "--output", str(out)])
            progress.update(1)

        doclinks_dir = out / "doclinks" / DOC_CATEGORY_DIR
        doclinks_dir.mkdir(parents=True, exist_ok=True)
        doc_id = uuid4().hex
        shutil.copy(out / "LOTO_A.pdf", doclinks_dir / f"{doc_id}.pdf")
        shutil.copy(out / "LOTO_A.json", doclinks_dir / f"{doc_id}.json")

        typer.echo(f"âœ… PDF + JSON saved to {out}")
        if open_pdf:
            pdf_path = out / "LOTO_A.pdf"
            try:
                typer.launch(str(pdf_path))
            except Exception:  # pragma: no cover - opening may fail in CI
                pass
    except Exception as exc:  # pragma: no cover - ensure friendly error
        typer.secho(f"Error: {exc}", err=True)
        raise typer.Exit(1)


if __name__ == "__main__":
    cli()
