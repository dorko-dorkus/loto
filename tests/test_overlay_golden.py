from __future__ import annotations

import json
from pathlib import Path

import pytest

from loto.models import IsolationAction, IsolationPlan
from loto.pid import build_overlay


@pytest.fixture
def golden(request):
    path = Path(request.node.fspath).with_suffix(".golden.json")
    expected = json.loads(path.read_text())

    def check(data: object) -> None:
        assert data == expected

    return check


def test_overlay_golden(golden) -> None:
    """Ensure overlay generation matches expected demo output."""

    map_path = Path(__file__).resolve().parent.parent / "demo/pids/pid_map.yaml"

    plan = IsolationPlan(
        plan_id="P1",
        actions=[
            IsolationAction(component_id="process:src->V-101", method="lock"),
            IsolationAction(component_id="process:src->V-102", method="lock"),
        ],
    )

    overlay = build_overlay(
        sources=["src"],
        asset="A-100",
        plan=plan,
        sim_fail_paths=[["src", "V-101", "A-100"]],
        map_path=map_path,
    )

    golden(overlay)
