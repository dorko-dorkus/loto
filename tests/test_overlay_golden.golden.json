{
  "highlight": [
    "#A100",
    "#SRC",
    "#V101",
    "#V102A",
    "#V102B"
  ],
  "badges": [
    {"selector": "#A100", "type": "asset"},
    {"selector": "#SRC", "type": "source"}
  ],
  "paths": [
    {"id": "path0", "selectors": ["#SRC", "#V101", "#A100"]}
  ],
  "warnings": []
}
